README.md
# Sensu Go Hangouts Chat Handler
![Go Test](https://github.com/betorvs/sensu-hangouts-chat-handler/workflows/Go%20Test/badge.svg)
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/betorvs/sensu-hangouts-chat-handler)

The Sensu Go Hangouts Chat Handler is a [Sensu Event Handler][3] which manages
[Hangouts Chat][2] for alerting purposes. With this handler,
[Sensu][1] can alert systems administrators in Hangouts Chats.

This handler was inspired by [hangouts ruby plugin][5].

## Installation

Download the latest version of the sensu-hangouts-chat-handler from [releases][4],
or create an executable script from this source.

From the local path of the sensu-hangouts-chat-handler repository:
```
go build -o /usr/local/bin/sensu-hangouts-chat-handler main.go
```

## Configuration

Example Sensu Go handler definition:

```yml
type: Handler
api_version: core/v2
metadata:
  name: hangouts-chat
  namespace: default
spec:
  type: pipe
  command: sensu-hangouts-chat-handler
  env_vars:
  - HANGOUTSCHAT_WEBHOOK="https://...."
  timeout: 10
  runtime_assets:
  - betorvs/sensu-hangouts-chat-handler
  filters:
  - is_incident
```

Example Sensu Go check definition:

```yml
---
type: CheckConfig
api_version: core/v2
metadata:
  name: dummy-app-healthz
  namespace: default
  annotations:
    documentation: https://docs.sensu.io/sensu-go/latest
spec:
  command: check-http -u http://localhost:8080/healthz
  subscriptions:
  - dummy
  handlers:
  - hangouts-chat
  interval: 60
  publish: true
```


## Usage Examples

Help:
```

The Sensu Go Google Hangsout handler for alerting

Usage:
  sensu-hangouts-chat-handler [flags]
  sensu-hangouts-chat-handler [command]

Available Commands:
  help        Help about any command
  version     Print the version number of this plugin

Flags:
  -A, --annotations-as-link string          Parse Check.metadata.annotations as link to post in Hangouts Chat. e. prometheus_url
  -E, --annotations-suffix-exclude string   Parse Check.metadata.annotations as link to post in Hangouts Chat. e. prometheus_url
  -S, --annotations-suffix-link string      Parse Check.metadata.annotations as link to post in Hangouts Chat. e. prometheus_url
  -L, --descriptionLimit int                The maximum length of the description field (default 1500)
  -d, --descriptionTemplate string          The template for the description to be sent (default "{{.Check.Output}}")
  -h, --help                                help for sensu-hangouts-chat-handler
  -l, --messageLimit int                    The maximum length of the message field (default 130)
  -m, --messageTemplate string              The template for the message to be sent (default "{{.Entity.Name}}/{{.Check.Name}}")
  -s, --sensuDashboard string               The HANGOUTS Chat Handler will use it to create a source Sensu Dashboard URL. Use HANGOUTSCHAT_SENSU_DASHBOARD. Example: http://sensu-dashboard.example.local/c/~/n (default "disabled")
      --threadKey                           Send message with &threadKey=event.id
      --threadKey-value string              Send message with &threadKey=event.id. Options: id or alias (entity.name-check.name) (default "id")
  -T, --titlePrettify                       Remove all -, /, \ and apply strings.Title in message title
  -w, --webhook string                      The Webhook URL, use default from HANGOUTSCHAT_WEBHOOK env var
  -a, --withAnnotations                     Include the event.metadata.Annotations in details to send to Hangouts Chat
  -W, --withLabels                          Include the event.metadata.Labels in details to send to Hangouts Chat

Use "sensu-hangouts-chat-handler [command] --help" for more information about a command.


```

**Note:** Make sure to set the `HANGOUTSCHAT_WEBHOOK` environment variable for sensitive credentials in production to prevent leaking into system process table. Please remember command arguments can be viewed by unprivileged users using commands such as `ps` or `top`. The `--webhook` argument is provided as an override primarily for testing purposes. 

### Argument Annotations

All arguments for this handler are tunable on a per entity or check basis based on annotations.  The
annotations keyspace for this handler is `sensu.io/plugins/sensu-hangouts-chat-handler/config`. 

#### Examples

To change the team argument for a particular check, for that checks's metadata add the following:

```yml
type: CheckConfig
api_version: core/v2
metadata:
  annotations:
    sensu.io/plugins/sensu-hangouts-chat-handler/config/webhook: "https://LongWebhookURLHere"
[...]
```


### Asset creation

The easiest way to get this handler added to your Sensu environment, is to add it as an asset from Bonsai:

```sh
sensuctl asset add betorvs/sensu-hangouts-chat-handler --rename sensu-hangouts-chat-handler
```

See `sensuctl asset --help` for details on how to specify version.

### troubleshoot tips

From [Sensu Docs][4]:
```bash
sensuctl event info <entity_name> <check_name> --format json |./sensu-hangouts-chat-handler
```

## Contributing

See https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md

[1]: https://github.com/sensu/sensu-go
[2]: https://developers.google.com/hangouts/chat
[3]: https://docs.sensu.io/sensu-go/5.0/reference/handlers/#how-do-sensu-handlers-work
[4]: https://github.com/betorvs/sensu-hangouts-chat-handler/releases
[5]: https://github.com/clevertoday/sensu-plugins-hangouts-chat
[6]: https://docs.sensu.io/sensu-go/latest/operations/maintain-sensu/troubleshoot/

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.2] - 2021-05-20
### Changed
- Add a checkURL before creating annotations. Inspired by [stackoverflow answer](https://stackoverflow.com/questions/31480710/validate-url-with-standard-package-in-go)

## [0.1.1] - 2021-04-29
### Added 
- flag `--threadKey-value` to switch between event.id and entity.name-check.name. 

### Changed
- flag `--enable-threadKey` to `--threadKey` and default value using event.id.
- remove `-race` flag from tests.
- update go. files

## [0.1.0] - 2021-04-28
### Added 
- flag `--enable-threadKey` to send message using `&threadKey=%s` equal entity.name-check.name. Then it keeps it in the same thread when it open and close an alert.

### Changed
- golang version to 1.16

## [0.0.11] - 2021-03-19
### Added
- flag `--titlePrettify` to apply strings.Title in message title and remove these characters -, /, \

### Changed
- update golang to 1.15

## [0.0.10] - 2021-12-21
### Added
- flag to parse annotations using strings.HasSuffix to create button: -S or --annotations-suffix-link
- flag to exclude annotations when creating buttons: -E or --annotations-suffix-exclude

## [0.0.9] - 2020-11-30
### Added
- Annotations and labels parsing checking

## [0.0.8] - 2020-11-30
### Added
- Add `--annotations-as-link` flag to parse annotation as link

## [0.0.7] - 2020-11-20
### Changed
- Fix bonsai asset list


## [0.0.6] - 2020-11-20
### Changed
- Changed go to 1.14

### Removed
- Removed travis-ci integration

### Added
- Added tests
- Added golanglint


## [0.0.5] - 2020-11-20
### Changed
- Changed webhook URL environment variable from `WEBHOOK_HANGOUTSCHAT` to `HANGOUTSCHAT_WEBHOOK`.
- Changed from spf13/cobra to sensu-community/sensu-plugin-sdk. 
- Changed `--withAnnotations` to parse all annotations, and exclude if it contains `sensu.io/plugins/sensu-hangouts-chat-handler/config`, and send as text to Hangouts Chat. 

### Added
- Added `--withLabels` to parse all labels, and exclude if it contains `sensu.io/plugins/sensu-hangouts-chat-handler/config`, and send as text to Hangouts Chat.

### Removed 
- Removed annotations as link in message.

## [0.0.4] - 2020-02-25
### Changed
- Change from text message to card message in Hangouts Chat

## [0.0.3] - 2020-01-16
### Changed
- Change from dep to go mod.
- gometalinter to golangci-lint
- correct goreleaser

## [0.0.2] - 2019-11-24
### Added
- Add `HANGOUTS_ANNOTATIONS` to parse annotations to include that information inside the alert.
### Changed
- Update README.


## [0.0.1] - 2019-10-30

### Added
- Initial release

main_test.go
package main

import (
	"encoding/json"
	"testing"

	"github.com/sensu/sensu-go/types"
	"github.com/stretchr/testify/assert"
)

func TestParseEventKeyTags(t *testing.T) {
	event := types.FixtureEvent("foo", "bar")
	_, err := json.Marshal(event)
	assert.NoError(t, err)
	plugin.MessageTemplate = "{{.Entity.Name}}/{{.Check.Name}}"
	plugin.MessageLimit = 100
	title := parseEventTitle(event)
	assert.Contains(t, title, "foo")
}

func TestParseDescription(t *testing.T) {
	event := types.FixtureEvent("foo", "bar")
	event.Check.Output = "Check OK"
	_, err := json.Marshal(event)
	assert.NoError(t, err)
	plugin.DescriptionTemplate = "{{.Check.Output}}"
	plugin.DescriptionLimit = 100
	description := parseDescription(event)
	assert.Equal(t, description, "Check OK")
}

func TestCheckArgs(t *testing.T) {
	assert := assert.New(t)
	event := types.FixtureEvent("entity1", "check1")
	assert.Error(checkArgs(event))
	plugin.Webhook = "https://longwebhook.com"
	assert.NoError(checkArgs(event))
}

func TestAnnotationsSlice(t *testing.T) {
	expectedTags := []string{"runbook_url"}
	plugin.AnnotationsAsLink = "runbook_url"
	tags := annotationsSlice()
	assert.Equal(t, tags, expectedTags)
	expectedTags2 := []string{"runbook_url", "docs_url"}
	plugin.AnnotationsAsLink = "runbook_url,docs_url"
	tags2 := annotationsSlice()
	assert.Equal(t, tags2, expectedTags2)
}

func TestStringInSlice(t *testing.T) {
	testSlice := []string{"foo", "bar", "test"}
	testString := "test"
	testResult := stringInSlice(testString, testSlice)
	assert.True(t, testResult)
}

func TestValidateDescription(t *testing.T) {
	event := types.FixtureEvent("entity1", "check1")
	event.Check.Annotations["runbook_url"] = "https://play.golang.org"
	event.Check.Annotations["sensu.io/plugins/sensu-hangouts-chat-handler/config/webhook"] = "https://LongWebhookURLHere.com"
	plugin.AnnotationsAsLink = "runbook_url"
	test1 := validateDescription("runbook_url")
	assert.Equal(t, test1, false)
	test2 := validateDescription("sensu.io/plugins/sensu-hangouts-chat-handler/config/webhook")
	assert.Equal(t, test2, false)
}

func TestEventDescription(t *testing.T) {
	event := types.FixtureEvent("entity1", "check1")
	event.Check.Annotations["runbook_url"] = "https://play.golang.org"
	plugin.AnnotationsAsLink = "runbook_url"
	test1 := eventDescription(event)
	assert.Contains(t, test1, "check1")
	assert.NotContains(t, test1, "check_runbook_url")
}

func TestTrim(t *testing.T) {
	testString := "This string is 33 characters long"
	assert.Equal(t, trim(testString, 40), testString)
	assert.Equal(t, trim(testString, 4), "This")
}

func TestAnnotationsExcludeSlice(t *testing.T) {
	expectedTags := []string{"runbook_url"}
	plugin.AnnotationsSuffixExcludeList = "runbook_url"
	tags := annotationsExcludeSlice()
	assert.Equal(t, tags, expectedTags)
	expectedTags2 := []string{"runbook_url", "docs_url"}
	plugin.AnnotationsSuffixExcludeList = "runbook_url,docs_url"
	tags2 := annotationsExcludeSlice()
	assert.Equal(t, tags2, expectedTags2)
}

func TestTitlePrettify(t *testing.T) {
	test1 := "long-check-with-too-many-dashes"
	res1 := titlePrettify(test1)
	val1 := "Long Check With Too Many Dashes"
	assert.Equal(t, val1, res1)
	test2 := "long-check-with-too-many-dashes/and/slashes-and\\others"
	res2 := titlePrettify(test2)
	val2 := "Long Check With Too Many Dashes And Slashes And Others"
	assert.Equal(t, val2, res2)
}

func TestCheckURL(t *testing.T) {
	test1 := checkURL("http:::/not.valid/a//a??a?b=&&c#hi")
	assert.False(t, test1)
	test2 := checkURL("http//google.com")
	assert.False(t, test2)
	test3 := checkURL("google.com")
	assert.False(t, test3)
	test4 := checkURL("/foo/bar")
	assert.False(t, test4)
	test5 := checkURL("http://google.com")
	assert.True(t, test5)
	test6 := checkURL("http://")
	assert.False(t, test6)
}


main.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/sensu-community/sensu-plugin-sdk/sensu"
	"github.com/sensu-community/sensu-plugin-sdk/templates"
	"github.com/sensu/sensu-go/types"
)

// Config represents the handler plugin config.
type Config struct {
	sensu.PluginConfig
	Webhook                      string
	SensuDashboard               string
	WithAnnotations              bool
	WithLabels                   bool
	MessageTemplate              string
	MessageLimit                 int
	DescriptionTemplate          string
	DescriptionLimit             int
	TitlePrettify                bool
	AnnotationsAsLink            string
	AnnotationsSuffixAsLink      string
	AnnotationsSuffixExcludeList string
	EnableThreadKey              bool
	ThreadKeyValue               string
}

var (
	plugin = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-hangouts-chat-handler",
			Short:    "The Sensu Go Google Hangsout handler for alerting",
			Keyspace: "sensu.io/plugins/sensu-hangouts-chat-handler/config",
		},
	}

	options = []*sensu.PluginConfigOption{
		{
			Path:      "webhook",
			Env:       "HANGOUTSCHAT_WEBHOOK",
			Argument:  "webhook",
			Shorthand: "w",
			Default:   "",
			Usage:     "The Webhook URL, use default from HANGOUTSCHAT_WEBHOOK env var",
			Value:     &plugin.Webhook,
		},
		{
			Path:      "sensuDashboard",
			Env:       "HANGOUTSCHAT_SENSU_DASHBOARD",
			Argument:  "sensuDashboard",
			Shorthand: "s",
			Default:   "disabled",
			Usage:     "The HANGOUTS Chat Handler will use it to create a source Sensu Dashboard URL. Use HANGOUTSCHAT_SENSU_DASHBOARD. Example: http://sensu-dashboard.example.local/c/~/n",
			Value:     &plugin.SensuDashboard,
		},
		{
			Path:      "withAnnotations",
			Env:       "",
			Argument:  "withAnnotations",
			Shorthand: "a",
			Default:   false,
			Usage:     "Include the event.metadata.Annotations in details to send to Hangouts Chat",
			Value:     &plugin.WithAnnotations,
		},
		{
			Path:      "withLabels",
			Env:       "",
			Argument:  "withLabels",
			Shorthand: "W",
			Default:   false,
			Usage:     "Include the event.metadata.Labels in details to send to Hangouts Chat",
			Value:     &plugin.WithLabels,
		},
		{
			Path:      "messageTemplate",
			Env:       "HANGOUTSCHAT_MESSAGE_TEMPLATE",
			Argument:  "messageTemplate",
			Shorthand: "m",
			Default:   "{{.Entity.Name}}/{{.Check.Name}}",
			Usage:     "The template for the message to be sent",
			Value:     &plugin.MessageTemplate,
		},
		{
			Path:      "messageLimit",
			Env:       "HANGOUTSCHAT_MESSAGE_LIMIT",
			Argument:  "messageLimit",
			Shorthand: "l",
			Default:   130,
			Usage:     "The maximum length of the message field",
			Value:     &plugin.MessageLimit,
		},
		{
			Path:      "descriptionTemplate",
			Env:       "HANGOUTSCHAT_DESCRIPTION_TEMPLATE",
			Argument:  "descriptionTemplate",
			Shorthand: "d",
			Default:   "{{.Check.Output}}",
			Usage:     "The template for the description to be sent",
			Value:     &plugin.DescriptionTemplate,
		},
		{
			Path:      "descriptionLimit",
			Env:       "HANGOUTSCHAT_DESCRIPTION_LIMIT",
			Argument:  "descriptionLimit",
			Shorthand: "L",
			Default:   1500,
			Usage:     "The maximum length of the description field",
			Value:     &plugin.DescriptionLimit,
		},
		{
			Path:      "titlePrettify",
			Env:       "",
			Argument:  "titlePrettify",
			Shorthand: "T",
			Default:   false,
			Usage:     "Remove all -, /, \\ and apply strings.Title in message title",
			Value:     &plugin.TitlePrettify,
		},
		{
			Path:      "annotations-as-link",
			Env:       "HANGOUTSCHAT_ANNOTATIONS_LINK",
			Argument:  "annotations-as-link",
			Shorthand: "A",
			Default:   "",
			Usage:     "Parse Check.metadata.annotations as link to post in Hangouts Chat. e. prometheus_url",
			Value:     &plugin.AnnotationsAsLink,
		},
		{
			Path:      "annotations-suffix-link",
			Env:       "HANGOUTSCHAT_ANNOTATIONS_SUFFIX_LINK",
			Argument:  "annotations-suffix-link",
			Shorthand: "S",
			Default:   "",
			Usage:     "Parse Check.metadata.annotations as link to post in Hangouts Chat. e. prometheus_url",
			Value:     &plugin.AnnotationsSuffixAsLink,
		},
		{
			Path:      "annotations-suffix-exclude",
			Env:       "HANGOUTSCHAT_ANNOTATIONS_SUFFIX_EXCLUDE",
			Argument:  "annotations-suffix-exclude",
			Shorthand: "E",
			Default:   "",
			Usage:     "Parse Check.metadata.annotations as link to post in Hangouts Chat. e. prometheus_url",
			Value:     &plugin.AnnotationsSuffixExcludeList,
		},
		{
			Path:      "threadKey",
			Env:       "",
			Argument:  "threadKey",
			Shorthand: "",
			Default:   false,
			Usage:     "Send message with &threadKey=event.id",
			Value:     &plugin.EnableThreadKey,
		},
		{
			Path:      "threadKey-value",
			Env:       "",
			Argument:  "threadKey-value",
			Shorthand: "",
			Default:   "id",
			Usage:     "Send message with &threadKey=event.id. Options: id or alias (entity.name-check.name)",
			Value:     &plugin.ThreadKeyValue,
		},
	}
)

// SliceCard struct
type SliceCard struct {
	Cards []Cards `json:"cards"`
}

// Header struct
type Header struct {
	Title    string `json:"title"`
	Subtitle string `json:"subtitle"`
	ImageURL string `json:"imageUrl"`
}

// KeyValue struct
type KeyValue struct {
	TopLabel         string `json:"topLabel"`
	Content          string `json:"content"`
	ContentMultiline bool   `json:"contentMultiline"`
}

// OpenLink struct
type OpenLink struct {
	URL string `json:"url"`
}

// OnClick struct
type OnClick struct {
	OpenLink OpenLink `json:"openLink"`
}

// TextButton struct
type TextButton struct {
	Text    string  `json:"text"`
	OnClick OnClick `json:"onClick"`
}

// Buttons struct
type Buttons struct {
	TextButton TextButton `json:"textButton"`
}

// Widgets struct
type Widgets struct {
	KeyValue *KeyValue `json:"keyValue,omitempty"`
	Buttons  []Buttons `json:"buttons,omitempty"`
}

// Sections struct
type Sections struct {
	Widgets []Widgets `json:"widgets"`
}

// Cards struct
type Cards struct {
	Header   Header     `json:"header"`
	Sections []Sections `json:"sections"`
}

// parseDescription func returns string with custom template string to use in description
func parseDescription(event *types.Event) (description string) {
	description, err := templates.EvalTemplate("description", plugin.DescriptionTemplate, event)
	if err != nil {
		return ""
	}
	// allow newlines to get expanded
	description = strings.Replace(description, `\n`, "\n", -1)
	return trim(description, plugin.DescriptionLimit)
}

// parseEventTitle func returns string
func parseEventTitle(event *types.Event) (title string) {
	title, err := templates.EvalTemplate("title", plugin.MessageTemplate, event)
	if err != nil {
		return ""
	}
	if plugin.TitlePrettify {
		newTitle := titlePrettify(title)
		return trim(newTitle, plugin.MessageLimit)
	}
	return trim(title, plugin.MessageLimit)
}

func main() {
	handler := sensu.NewGoHandler(&plugin.PluginConfig, options, checkArgs, executeHandler)
	handler.Execute()
}

func checkArgs(_ *types.Event) error {
	if len(plugin.Webhook) == 0 {
		return fmt.Errorf("webhook url for Hangsout Chat is empty")
	}
	return nil
}

// formattedEventAction func
func formattedEventAction(event *types.Event) string {
	switch event.Check.Status {
	case 0:
		return "RESOLVED"
	default:
		return "ALERT"
	}
}

// parseAnnotationsToButton func
func parseAnnotationsToButton(event *types.Event) []Buttons {
	var button []Buttons

	if plugin.SensuDashboard != "disabled" && checkURL(plugin.SensuDashboard) {
		newbutton := Buttons{}
		newbutton.TextButton.Text = "Sensu Source"
		newbutton.TextButton.OnClick.OpenLink.URL = fmt.Sprintf("%s/%s/events/%s/%s", plugin.SensuDashboard, event.Entity.Namespace, event.Entity.Name, event.Check.Name)
		button = append(button, newbutton)
	}
	if plugin.AnnotationsAsLink != "" {
		tags := annotationsSlice()
		if event.Check.Annotations != nil {
			for key, value := range event.Check.Annotations {
				if stringInSlice(key, tags) && checkURL(value) {
					newbutton := Buttons{}
					newbutton.TextButton.Text = key
					newbutton.TextButton.OnClick.OpenLink.URL = value
					button = append(button, newbutton)
				}
			}
		}
	}
	if plugin.AnnotationsSuffixAsLink != "" {
		// tags := annotationsSlice()
		if event.Check.Annotations != nil {
			for key, value := range event.Check.Annotations {
				if plugin.AnnotationsSuffixExcludeList != "" {
					tags := annotationsExcludeSlice()
					if stringInSlice(key, tags) {
						continue
					}
				}
				// && strings.Contains(key, )
				if strings.HasSuffix(key, plugin.AnnotationsSuffixAsLink) && checkURL(value) {

					newbutton := Buttons{}
					newbutton.TextButton.Text = key
					newbutton.TextButton.OnClick.OpenLink.URL = value
					button = append(button, newbutton)
				}

			}
		}
	}
	if len(button) == 0 {
		newbutton := Buttons{}
		newbutton.TextButton.Text = "Sensu Documentation"
		newbutton.TextButton.OnClick.OpenLink.URL = "https://docs.sensu.io/sensu-go/latest"
		button = append(button, newbutton)
	}
	return button
}

// // eventDescription func return an message to use it
func eventDescription(event *types.Event) string {
	var (
		annotations string
		labels      string
		message     string
	)
	message = fmt.Sprintf("Entity: %s, \nCheck: %s, \nCommand: %s, ", event.Entity.Name, event.Check.Name, event.Check.Command)
	if event.Check.ProxyEntityName != "" {
		message += fmt.Sprintf("\nProxy_Entity: %s, \n", event.Check.ProxyEntityName)
	}
	if plugin.WithAnnotations {
		if event.Annotations != nil {
			for key, value := range event.Annotations {
				if validateDescription(key) {
					annotations += fmt.Sprintf("%s_%s: %s, \n", "event", key, value)
				}
			}
		}
		if event.Check.Annotations != nil {
			for key, value := range event.Check.Annotations {
				if validateDescription(key) {
					annotations += fmt.Sprintf("%s_%s: %s, \n", "check", key, value)
				}
			}
		}
		if event.Entity.Annotations != nil {
			for key, value := range event.Check.Annotations {
				if validateDescription(key) {
					annotations += fmt.Sprintf("%s_%s: %s, \n", "entity", key, value)
				}
			}
		}
		if annotations != "" {
			message += fmt.Sprintf("\n Annotations: \n%s", annotations)
		}
	}
	if plugin.WithLabels {
		if event.Labels != nil {
			for key, value := range event.Labels {
				if validateDescription(key) {
					labels += fmt.Sprintf("%s_%s: %s, \n", "event", key, value)
				}
			}
		}
		if event.Check.Labels != nil {
			for key, value := range event.Check.Labels {
				if validateDescription(key) {
					labels += fmt.Sprintf("%s_%s: %s, \n", "check", key, value)
				}
			}
		}
		if event.Entity.Labels != nil {
			for key, value := range event.Check.Labels {
				if validateDescription(key) {
					labels += fmt.Sprintf("%s_%s: %s, \n", "entity", key, value)
				}
			}
		}
		if labels != "" {
			message += fmt.Sprintf("\n Labels: \n%s", labels)
		}
	}
	return message
}

func validateDescription(key string) bool {
	// tags := []string{plugin.Keyspace}
	if plugin.AnnotationsAsLink != "" {
		tags := annotationsSlice()
		if !stringInSlice(key, tags) && !strings.Contains(key, plugin.Keyspace) {
			return true
		}
	}

	return false
}

// run func do everything
func executeHandler(event *types.Event) error {

	keyvalue1 := KeyValue{
		TopLabel:         formattedEventAction(event),
		Content:          eventDescription(event),
		ContentMultiline: true,
	}
	// keyvalue2 := KeyValue{
	// 	TopLabel: "More Information",
	// 	Content:  parseAnnotations(event),
	// }
	keyvalue3 := KeyValue{
		TopLabel:         "Check Output",
		Content:          parseDescription(event),
		ContentMultiline: true,
	}
	widget1 := Widgets{
		KeyValue: &keyvalue1,
	}
	widget2 := Widgets{
		// KeyValue: &keyvalue2,
		Buttons: parseAnnotationsToButton(event),
	}
	widget3 := Widgets{
		KeyValue: &keyvalue3,
	}
	header := Header{
		Title:    "Sensu Event",
		Subtitle: parseEventTitle(event),
	}
	section1 := Sections{
		Widgets: []Widgets{widget1},
	}
	section2 := Sections{
		Widgets: []Widgets{widget2},
	}
	section3 := Sections{
		Widgets: []Widgets{widget3},
	}
	card := Cards{
		Header:   header,
		Sections: []Sections{section1, section2, section3},
	}
	formPost := SliceCard{
		Cards: []Cards{card},
	}
	// prettyJSON, err := json.MarshalIndent(formPost, "", "  ")
	// if err != nil {
	// 	fmt.Printf("Failed to generate json %v", err)
	// }
	// fmt.Printf("%s\n", string(prettyJSON))
	bodymarshal, err := json.Marshal(&formPost)
	if err != nil {
		fmt.Printf("[ERROR] %s", err)
	}
	webhook := plugin.Webhook
	if plugin.EnableThreadKey {
		switch plugin.ThreadKeyValue {
		case "id":
			webhook = fmt.Sprintf("%s&threadKey=%s", plugin.Webhook, event.GetUUID())
		case "alias":
			webhook = fmt.Sprintf("%s&threadKey=%s", plugin.Webhook, parseAlias(event))
		default:
			webhook = fmt.Sprintf("%s&threadKey=%s", plugin.Webhook, event.GetUUID())
		}
	}
	err = Post(webhook, bodymarshal)
	if err != nil {
		fmt.Printf("[ERROR] %s", err)
	}
	return nil
}

//Post func to send the json to hangouts chat
func Post(url string, body []byte) error {
	client := &http.Client{
		Timeout: time.Second * 10,
	}
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(body))
	if err != nil {
		fmt.Printf("[ERROR] %s", err)
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("[ERROR] %s", err)
	}
	if resp.StatusCode != 200 {
		bodyText, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			fmt.Printf("[ERROR] %s", err)
		}
		s := string(bodyText)
		fmt.Printf("[LOG]: %s ; %s", resp.Status, s)
	}
	defer resp.Body.Close()
	return nil
}

func annotationsSlice() []string {
	tags := []string{}
	if strings.Contains(plugin.AnnotationsAsLink, ",") {
		tags = strings.Split(plugin.AnnotationsAsLink, ",")
	}
	if !strings.Contains(plugin.AnnotationsAsLink, ",") {
		tags = append(tags, plugin.AnnotationsAsLink)
	}

	return tags
}

func annotationsExcludeSlice() []string {
	tags := []string{}
	if strings.Contains(plugin.AnnotationsSuffixExcludeList, ",") {
		tags = strings.Split(plugin.AnnotationsSuffixExcludeList, ",")
	}
	if !strings.Contains(plugin.AnnotationsSuffixExcludeList, ",") {
		tags = append(tags, plugin.AnnotationsSuffixExcludeList)
	}

	return tags
}

// use to parse annotations to send as link
func stringInSlice(a string, list []string) bool {
	for _, b := range list {
		if b == a {
			return true
		}
	}
	return false
}

// time func returns only the first n bytes of a string
func trim(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

func titlePrettify(s string) string {
	var title string
	title = strings.ReplaceAll(s, "-", " ")
	title = strings.ReplaceAll(title, "\\", " ")
	title = strings.ReplaceAll(title, "/", " ")
	title = strings.Title(title)
	return title
}

func parseAlias(event *types.Event) string {
	return fmt.Sprintf("%s-%s", event.Entity.Name, event.Check.Name)
}

func checkURL(str string) bool {
	u, err := url.Parse(str)
	return err == nil && u.Scheme != "" && u.Host != ""
}
